File: Day 19/Part_One.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long int
const int MOD =  1e9 + 7;
const int N = 1e5 + 5;
const int INF = 1e10;

void compute()
{

}

void solve()
{   
    int n; cin >> n;
    int ans = 0;

    vector <pair<int, int>> points;
    points.push_back({0, 0});
    int x = 0, y = 0;

    map <char, pair<int, int>> mp;
    mp['R'] = {0, 1};
    mp['L'] = {0, -1};
    mp['U'] = {1, 0};
    mp['D'] = {-1, 0};

    int perimeter = 0;

    for(int i=0; i<n; i++)
    {
        char dir; cin >> dir;
        int temp; cin >> temp;
        string color; cin >> color;

        perimeter += temp;

        x += mp[dir].first * temp;
        y += mp[dir].second * temp;

        points.push_back({x, y});
    }

    // Shoelace Theorem
    int area = 0;
    for(int i=1; i<points.size(); i++)
    {
        area += ((points[i].first * points[(i+1)%(n+1)].second) - (points[i].second * points[(i+1)%(n+1)].first));
    }
  
    // Pick's Theorem
    // A = i + b/2 - 1
    // i = A - b/2 + 1

    // ans = i + b = A + b/2 + 1
    area = area*0.5;
    cout << area + perimeter + 1 << '\n';
}

signed main(void) 
{
#ifndef ONLINE_JUDGE
    freopen("input.txt","r",stdin);
    freopen("output.txt","w",stdout);
#endif
 
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int t = 1;
    // cin >> t;

    compute();
    
    for(int i=0;i<t;i++)
    {
        // cout << "Case #" << i+1 << ": "; // Kick Start
        solve();
    }

	return 0;
}

File: Day 19/Part_One.py
import sys
sys.stdin = open("input.txt", "r")
sys.stdout = open("output.txt", "w")
sys.stderr = open("output.txt", "w")

with open("input.txt", "r") as f:
    data = f.read().split("\n\n")
    
    compute = data[0].split()
    query = data[1].split()

mp = {}
for i in compute:
    mp[i.split("{")[0]] = i.split("{")[1][:-1].split(',')

ans = 0
for i in query:
    val = dict([j.split(":") for j in i.replace("=", ":")[1:-1].split(",")])
    val = dict([[j, int(val[j])] for j in val])
    
    start = "in"
    flag = 0

    while(True):

        if(start == 'A'):
            flag = 1
            break

        if(start == 'R'):
            flag = 0
            break

        l = mp[start]
        for j in l:
            if(j.isalpha()):
                start = j
                break
            else:
                temp = j.split(":")[0]
                next = j.split(":")[1]

                if(temp[1] == "<" and val[temp[0]] < int(temp[2:])):
                    start = next
                    break
                elif(temp[1] == ">" and val[temp[0]] > int(temp[2:])):
                    start = next
                    break

    if(flag == 1):
        ans += sum([val[j] for j in val])
                

print(ans)
